### PR TITLE
Union, Clone and Copy

### PR DESCRIPTION
**Pull request by tmiasko**
Functionally wise this PR introduces support for unions and derives Clone and
Copy traits for ffi types (based on @Luke-Nukem work from #415).

It also temporarily removes support for structs with partially generated
fields. In particular types with bitfields are not generated, with exception of
GdkEventKey and GdkEventScroll. Exception is made because there is manual code
that uses those.

From implementation perspective this change tries to migrate most logic to
phases that occur before code generation, so that codegen is only concerned
with joining some string together, so to say. At this point this may not be
readily visible, as codegen cleanup will follow in a later PR.

Tested with all crates hosted under gtk-rs (regenerated ffi and api crates).
soup-sys-rs requires manual tweak to import sockaddr from libc.

This PR incorporates both #415 and #526.

### CODE REVIEW

**Comment by EPashkin** [comment]
Seems blocked by https://github.com/gtk-rs/gir-files/issues/13 as attempt to generate `pangocairo` causes "thread 'main' has overflowed its stack" runtime error

**Comment by tmiasko** [comment]
@EPashkin thanks for pointing that out. I am running this with PangoCairo fixed
as descried in gtk-rs/gir-files#13, i.e., all fields are removed from FcFontMap.

**Comment by EPashkin** [comment]
Generate empty lines in gdk-sys
```diff
@@ -3210,6 +3229,7 @@ impl ::std::fmt::Debug for GdkEventGrabBroken {
 }
 
 #[repr(C)]
+
 pub struct GdkEventKey {
     pub type_: GdkEventType,
     pub window: *mut GdkWindow,
@@ -3354,6 +3381,7 @@ impl ::std::fmt::Debug for GdkEventProximity {
 }
 
 #[repr(C)]
+
 pub struct GdkEventScroll {
     pub type_: GdkEventType,
     pub window: *mut GdkWindow,
```

Examples builds fine but there warning about `impl ::std::fmt::Debug for GDoubleIEEE754 {` and `impl ::std::fmt::Debug for GFloatIEEE754`
```
    --> d:/eap/rust/0/sys/glib-sys\src\lib.rs:1121:9
     |
1121 |         unsafe {
     |         ^^^^^^ unnecessary `unsafe` block
     |
     = note: #[warn(unused_unsafe)] on by default

warning: unnecessary `unsafe` block
    --> d:/eap/rust/0/sys/glib-sys\src\lib.rs:1133:9
     |
1133 |         unsafe {
     |         ^^^^^^ unnecessary `unsafe` block
```
and `impl ::std::fmt::Debug for GdkEvent`
```
warning: unnecessary `unsafe` block
    --> d:/eap/rust/0/sys/gdk-sys\src\lib.rs:3028:9
     |
3028 |         unsafe {
     |         ^^^^^^ unnecessary `unsafe` block
     |
     = note: #[warn(unused_unsafe)] on by default
```

**Comment by EPashkin** [comment]
@tmiasko Thanks, I check it tomorrow.
Currently I only can say that it add too many unneeded empty lines

File: src/codegen/sys/fields.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        truncated,
+        derives_copy: union.derives_copy(&env.library),
+        kind: "union",
+        fields,
+    }
+}
+
+fn analyze_fields(env: &Env, unsafe_access: bool, fields: &[Field]) -> (Vec<FieldInfo>, Option<String>) {
+    let mut truncated = None;
+    let mut infos = Vec::new();
```
Comment (EPashkin) [inline src/codegen/sys/fields.rs]: Maybe use `Vec::with_capacity(fields.len)` ?

Comment (tmiasko) [reply src/codegen/sys/fields.rs]: Done.

File: src/codegen/sys/fields.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        });
+    }
+
+    (infos, truncated)
+}
+
+
+fn field_ffi_type(env: &Env, field: &Field) -> Result {
+    if field.is_incomplete(&env.library) {
+        return Err(TypeError::Ignored(format!("field {} has incomplete type", &field.name)));
```
Comment (EPashkin) [inline src/codegen/sys/fields.rs]: IMHO better also show type here (not insist, as do it troublesome at first glance).
```
#[repr(C)]
pub struct GtkAboutDialogClass {
    _truncated_record_marker: c_void,
    // /*Ignored*/field parent_class has incomplete type
}
```

File: src/analysis/types.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+/// Checks if given type is some kind of pointer.
+pub trait IsPtr {
+    fn is_ptr(&self) -> bool;
+}
+
+impl IsPtr for Field {
+    fn is_ptr(&self) -> bool {
+        if let Some(ref c_type) = self.c_type {
+            return c_type.contains("*")
```
Comment (EPashkin) [inline src/analysis/types.rs]: Clippy nags about this and others `contains("*")`:
```
warning: single-character string constant used as pattern
  --> src\analysis\types.rs:16:36
   |
16 |             return c_type.contains("*")
   |                    ----------------^^^- help: try using a char instead: `c_type.contains('*')`
   |
   = note: #[warn(single_char_pattern)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.183/index.html#single_char_pattern
```

File: src/library_postprocessing.rs [hunk 4]
```diff
@@ -268,4 +273,105 @@ impl Library {
...
+                        for (fid, field) in fields.iter().enumerate() {
+                            if nameutil::needs_mangling(&field.name) {
+                                let new_name = nameutil::mangle_keywords(&*field.name).to_string();
+                                actions.push((tid, fid, Action::SetName(new_name)));
+                            }
+                            if field.c_type.is_some() {
+                                continue;
+                            }
+                            let field_type = self.type_(field.typ);
+                            if let Some(_) = field_type.maybe_ref_as::<Function>() {
```
Comment (EPashkin) [inline src/library_postprocessing.rs]:
```
warning: redundant pattern matching, consider using `is_some()`
   --> src\library_postprocessing.rs:303:36
    |
303 |                             if let Some(_) = field_type.maybe_ref_as::<Function>() {
    |                             -------^^^^^^^---------------------------------------- help: try this: `if field_type.maybe_ref_as::<Function>().is_some()`
    |
    = note: #[warn(if_let_redundant_pattern_matching)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.183/index.html#if_let_redundant_pattern_matching
```

File: src/library_postprocessing.rs [hunk 5]
```diff
@@ -268,4 +273,105 @@ impl Library {
...
+                    ns_id: ns_id as u16,
+                    id: id as u32,
+                };
+                match *type_ {
+                    Type::Class(Class { ref name, ref fields, ..}) |
+                    Type::Record(Record { ref name, ref fields, ..}) |
+                    Type::Union(Union { ref name, ref fields, ..}) => {
+                        for (fid, field) in fields.iter().enumerate() {
+                            if nameutil::needs_mangling(&field.name) {
+                                let new_name = nameutil::mangle_keywords(&*field.name).to_string();
```
Comment (EPashkin) [inline src/library_postprocessing.rs]: Maybe  `.into_owned()` instead?

File: src/library_postprocessing.rs [hunk 6]
```diff
@@ -268,4 +273,105 @@ impl Library {
...
+                Type::Union(Union {ref name, ref mut fields, ..}) => {
+                    info!("Type `{}` is not representable.", name);
+                    fields.clear();
+                }
+                _ => unreachable!("Expected a union"),
+            }
+        }
+    }
 }
+
```
Comment (EPashkin) [inline src/library_postprocessing.rs]: Unneeded line

File: src/parser.rs [hunk 7]
```diff
@@ -285,14 +285,7 @@ impl Library {
...
-                            #[cfg(not(feature = "use_unions"))]
-                            {
-                                Type::union(self, u, INTERNAL_NAMESPACE)
-                            }
-                            #[cfg(feature = "use_unions")]
-                            {
-                                Type::union(self, u, ns_id)
-                            }
+                            Type::union(self, u, ns_id)
                         };
```
Comment (EPashkin) [inline src/parser.rs]: Just `let type_id = Type::union(self, u, ns_id)` ?

File: src/parser.rs [hunk 8]
```diff
@@ -436,14 +429,7 @@ impl Library {
...
                             let type_id = {
-                                #[cfg(not(feature = "use_unions"))]
-                                {
-                                    Type::union(self, u, INTERNAL_NAMESPACE)
-                                }
-                                #[cfg(feature = "use_unions")]
-                                {
-                                    Type::union(self, u, ns_id)
-                                }
+                                Type::union(self, u, ns_id)
```
Comment (EPashkin) [inline src/parser.rs]: Same

File: src/parser.rs [hunk 9]
```diff
@@ -591,7 +583,7 @@ impl Library {
                                         s.push('_');
                                         s
                                     })
-                                    .unwrap_or_else(String::new),
+                                    .unwrap_or("".into()),
```
Comment (EPashkin) [inline src/parser.rs]: Please change it back, and next too

Comment (EPashkin) [reply src/parser.rs]: .. as clippy nags

File: src/parser.rs (hunk 7 above)
Comment (EPashkin) [reply src/parser.rs]: Or move into `fields.push`

File: src/codegen/sys/statics.rs [hunk 10]
```diff
@@ -52,7 +47,7 @@ pub fn only_for_glib(w: &mut Write) -> Result<()> {
         "pub type gpointer = *mut c_void;",
         "",
         "#[repr(C)]",
-        "#[derive(Debug)]",
+        "#[derive(Copy,Clone,Debug)]",
```
Comment (EPashkin) [inline src/codegen/sys/statics.rs]: `#[derive(Copy, Clone, Debug)]` ?

File: src/analysis/types.rs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+impl ImplementsDebug for Field {
+    fn implements_debug(&self, lib: &Library) -> bool {
+        match *lib.type_(self.typ) {
+            Type::FixedArray(_, size) => size <= RUST_DERIVE_ARRAY_SIZE_LIMIT,
+            Type::Function(Function {ref parameters, ..}) => parameters.len() <= RUST_DERIVE_PARAM_SIZE_LIMIT,
+            _ => true,
+        }
+    }
+}
+
```
Comment (EPashkin) [inline src/analysis/types.rs]: Unneeded line

**Comment by EPashkin** [comment]
@tmiasko, @sdroege, @GuillaumeGomez There will be no problem with union Debug implementation?
I worry about cases when value unrepresentative on some fields,  but not sure that it can be happened.
```rust
#[repr(C)]
#[derive(Copy, Clone)]
pub union GtkBindingArg_d {
    pub long_data: c_long,
    pub double_data: c_double,
    pub string_data: *mut c_char,
}

impl ::std::fmt::Debug for GtkBindingArg_d {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        f.debug_struct(&format!("GtkBindingArg_d @ {:?}", self as *const _))
         .field("long_data", unsafe { &self.long_data })
         .field("double_data", unsafe { &self.double_data })
         .field("string_data", unsafe { &self.string_data })
         .finish()
    }
}
```

**Comment by EPashkin** [comment]
Also we need print `padding` values? (There can be needed "padding" fields)
```rust
#[repr(C)]
#[derive(Copy, Clone)]
pub union GtkTextAttributes_u1 {
    pub font_features: *mut c_char,
    pub padding: [c_uint; 2],
}

impl ::std::fmt::Debug for GtkTextAttributes_u1 {
    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
        f.debug_struct(&format!("GtkTextAttributes_u1 @ {:?}", self as *const _))
         .field("font_features", unsafe { &self.font_features })
         .field("padding", unsafe { &self.padding })
         .finish()
    }
}
```

**Comment by EPashkin** [comment]
Note: almost all GUI "xxClass" now incomplete due bit field in GtkContainerClass
```diff
@@ -3083,20 +3494,24 @@ pub struct GtkContainerClass {
     pub get_child_property: Option<unsafe extern "C" fn(*mut GtkContainer, *mut GtkWidget, c_uint, *mut gobject::GValue, *mut gobject::GParamSpec)>,
     pub get_path_for_child: Option<unsafe extern "C" fn(*mut GtkContainer, *mut GtkWidget) -> *mut GtkWidgetPath>,
     _truncated_record_marker: c_void,
-    //_handle_border_width: unsigned: 1,
-    //_gtk_reserved1: fn,
-    //_gtk_reserved2: fn,
-    //_gtk_reserved3: fn,
-    //_gtk_reserved4: fn,
-    //_gtk_reserved5: fn,
-    //_gtk_reserved6: fn,
-    //_gtk_reserved7: fn,
-    //_gtk_reserved8: fn,
+    // /*Ignored*/field _handle_border_width has incomplete type
 }
```

**Comment by tmiasko** [comment]
My thoughts about issues raised by @EPashkin, and some of my own:

* Should we ignore Debug impl for Unions? I start with assumption that Rust
  treats unions essentially as sequence of bytes (which is somewhat open
  question at the moment AFAIK). On that basis, it would seem that accessing
  union fields should be fine. We only have primitive types there (or other
  struct / union that have only primitive types), and pointers are NOT
  dereferenced.

* Should we ignore private fields in Debug? The padding fields should fall into
  this category. I think we should omit them from Debug impl and will update
  code latter accordingly.

* Should we generate private fields with private visibility, i.e., without "pub"?
  I think we should, but glib crate currently accesses those ...

Thanks for review @EPashkin.

**Comment by tmiasko** [comment]
I pushed two commits. First addresses @EPashkin review comments.
Second one omits private and volatile fields from Debug implementation. 

It did help to some degree with printing padding fields, 
but some are still left like in GSettingsClass.

**Comment by GuillaumeGomez** [comment]
The best way to check if `Debug` implementations are working is to actually test them. Like that, we'll know.

**Comment by EPashkin** [comment]
@GuillaumeGomez  I forgot that we can test it ;)
```
let u = gtk_ffi::GtkBindingArg_d { long_data: 0x12345678i32 };
println!("{:?}", u);
```
print
```
GtkBindingArg_d @ 0x58df2d8 { long_data: 305419896, double_data: 0.000000000000000000000000000000000000000000000000000000000000000000000000000
0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002272893269, string_data: 0x112345678 }
```

File: src/analysis/types.rs (hunk 11 above)
Comment (EPashkin) [reply src/analysis/types.rs]: This still actual

File: src/codegen/sys/statics.rs [hunk 12]
```diff
@@ -52,7 +47,7 @@ pub fn only_for_glib(w: &mut Write) -> Result<()> {
         "pub type gpointer = *mut c_void;",
         "",
         "#[repr(C)]",
-        "#[derive(Debug)]",
+        "#[derive(Copy,Clone)]",
```
Comment (EPashkin) [inline src/codegen/sys/statics.rs]: `"#[derive(Copy, Clone)]",` ?

Comment (tmiasko) [reply src/codegen/sys/statics.rs]: Union containing volatile field need to be copyable as well.
Alternatively we could stop generating unions with volatile fields.

**Comment by EPashkin** [comment]
Changes from last 2 commits:
```diff
@@ -1261,11 +1258,9 @@ pub struct GApplicationCommandLineClass {
 impl ::std::fmt::Debug for GApplicationCommandLineClass {
     fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
         f.debug_struct(&format!("GApplicationCommandLineClass @ {:?}", self as *const _))
-         .field("parent_class", &self.parent_class)
          .field("print_literal", &self.print_literal)
          .field("printerr_literal", &self.printerr_literal)
          .field("get_stdin", &self.get_stdin)
-         .field("padding", &self.padding)
          .finish()
     }
 }
@@ -1554,7 +1549,6 @@ pub struct GDBusAnnotationInfo {
 impl ::std::fmt::Debug for GDBusAnnotationInfo {
     fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
         f.debug_struct(&format!("GDBusAnnotationInfo @ {:?}", self as *const _))
-         .field("ref_count", &self.ref_count)
          .field("key", &self.key)
          .field("value", &self.value)
          .field("annotations", &self.annotations)
@@ -4168,8 +4145,6 @@ pub struct GSimpleActionGroupClass {
 impl ::std::fmt::Debug for GSimpleActionGroupClass {
     fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
         f.debug_struct(&format!("GSimpleActionGroupClass @ {:?}", self as *const _))
-         .field("parent_class", &self.parent_class)
-         .field("padding", &self.padding)
          .finish()
     }
 }
@@ -1144,8 +1144,6 @@ pub union GMutex {
 impl ::std::fmt::Debug for GMutex {
     fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
         f.debug_struct(&format!("GMutex @ {:?}", self as *const _))
-         .field("p", unsafe { &self.p })
-         .field("i", unsafe { &self.i })
          .finish()
     }
 }
```

File: src/codegen/sys/statics.rs (hunk 12 above)
Comment (EPashkin) [reply src/codegen/sys/statics.rs]: Sorry that it was unclear: You just forgot space between.

Comment (EPashkin) [reply src/codegen/sys/statics.rs]: Please make 2 last fixes change current commits.

**Comment by tmiasko** [comment]
Done. I also included hack for GHookList so that gstreamer bindings remain more or less working.

**Comment by EPashkin** [comment]
Thanks

**Comment by EPashkin** [comment]
Ok, no one against PR, so I will merge it.

**Pull request merged**